index.js

```diff
@@ -5,31 +5,95 @@ const yaml = require('js-yaml');
 const requestHandler = require('./lib/request-handler');
 const eventHandler = require('./lib/event-handler');
 const Context = require('./lib/context');
+const shutdown = require('death')({ uncaughtException: true });
 
 // HTTP framework
 const fastify = require('fastify');
 
-let LOG_LEVEL = 'warn';
+// Default log level
+const LOG_LEVEL = 'warn';
+
+// Default port
 const PORT = 8080;
 
-// Invoker
-function start(func, options) {
+/**
+ * Starts the provided Function. If the function is a module, it will be
+ * inspected for init, shutdown, liveness, and readiness functions and those
+ * will be used to configure the server. If it's a function, it will be used
+ * directly.
+ *
+ * @param {Object | function} func The function to start (see the Function type)
+ * @param {*} options Options to configure the server
+ * @param {string} options.logLevel The log level to use
+ * @param {number} options.port The port to listen on
+ * @returns {Promise<http.Server>} The server that was started
+ */
+async function start(func, options) {
   options = options || {};
+  if (typeof func === 'function') {
+    return __start(func, options);
+  }
+  if (typeof func.handle !== 'function') {
+    throw new TypeError('Function must export a handle function');
+  }
+  if (typeof func.init === 'function') {
+    func.init();
+  }
+  if (typeof func.shutdown === 'function') {
+    options.shutdown = func.shutdown;
+  }
+  if (typeof func.liveness === 'function') {
+    options.liveness = func.liveness;
+  }
+  if (typeof func.readiness === 'function') {
+    options.readiness = func.readiness;
+  }
+  return __start(func.handle, options);
+}
 
+/**
+ * Internal function to start the server. This is used by the start function.
+ *
+ * @param {function} func - The function to start
+ * @param {*} options - Options to configure the server
+ * @param {string} options.logLevel - The log level to use
+ * @param {number} options.port - The port to listen on
+ * @returns {Promise<http.Server>} The server that was started
+ */
+async function __start(func, options) {
   // Load a func.yaml file if it exists
-  const funcConfig = loadFuncYaml(options.config) || {};
+  const config = loadConfig(options);
 
-  // Set the log level
-  if (['fatal', 'error', 'warn', 'info', 'debug', 'trace', 'silent']
-    .find(l => funcConfig.logLevel === l)) {
-    LOG_LEVEL = funcConfig.logLevel;
-  }
+  // Create and configure the server for the default behavior
+  const server = initializeServer(config);
+
+  // Configures the server to handle incoming requests to the function itself,
+  // and also to other endpoints such as telemetry and liveness/readiness
+  requestHandler(server, { func, funcConfig: config });
+
+  // Start the server
+  return new Promise((resolve, reject) => {
```

Comment:
If the `__start()` function is `async` couldn't we just use
```js
await server.listen({
  port: config.port,
  host: '::'
}
```
instead of `new Promise()`?

Reply:
Or maybe it does support `await`.
https://www.fastify.io/docs/latest/Guides/Getting-Started/
```js
/**
 * Run the server!
 */
const start = async () => {
  try {
    await fastify.listen({ port: 3000 })
  } catch (err) {
    fastify.log.error(err)
    process.exit(1)
  }
}
start()
```